packages/runtime/src/controls/controls/shape.ts

```diff
@@ -0,0 +1,43 @@
+import { CopyContext } from '../../state/react-page'
+import { ControlDefinition, ControlDefinitionData } from '../control'
+
+import { copy as controlCopy } from '../copy'
+
+export const ShapeControlType = 'makeswift::controls::shape'
+
+type ShapeControlConfig = {
+  type: Record<string, ControlDefinition>
+}
+
+export type ShapeControlDefinition<C extends ShapeControlConfig = ShapeControlConfig> = {
+  type: typeof ShapeControlType
+  config: C
+}
+
+export function Shape<C extends ShapeControlConfig>(config: C): ShapeControlDefinition<C> {
+  return { type: ShapeControlType, config }
+}
+
+export type ShapeControlData<T extends ShapeControlDefinition = ShapeControlDefinition> = {
+  [K in keyof T['config']['type']]?: ControlDefinitionData<T['config']['type'][K]>
+}
+
+export function shapeCopy(
+  definition: ShapeControlDefinition,
+  value: ShapeControlData,
+  context: CopyContext,
+): ShapeControlData {
+  const newValue: ShapeControlData = {}
+
+  for (const [key, itemDefinition] of Object.entries(definition.config.type)) {
+    const prop = value[key]
+
+    // @note: This must call controlCopy directly, rather than the copyElement function
```

Comment:
I'm not sure I understand why this comment is necessary. Why would a `Shape` control need to call `copyElement` at all? It's just a composition of other controls, so it seems that it calling the function to copy control data should be obvious, no?

Reply:
There's an interesting parallel between the slot control and the shape control.

The slot control allows the builder user to compose element data; the shape control allows the developer to compose prop data.

Now, element data contains prop data, but not the other way around. This was not intuitive to me at first. I thought `Shape` would be similar to `Slot`, but I had to rely on the config. I can imagine a world where we put that config structure in the data itself — or more specifically, where we could derive a control definition from the data itself (that's what we do with `Slot`).

That wasn't clear to me, which is why I included the comment.

We can refine or remove it; my goal is to increase the understanding of the people reading this in the future

Reply:
Ok, cool. Yeah this distinction did confuse me at first, perhaps it still does. I'm sure we'll keep talking about it over time. Removed the comment